common_util/src/future_cancel.rs

```diff
@@ -0,0 +1,91 @@
+// Copyright 2022-2023 CeresDB Project Authors. Licensed under Apache-2.0.
+
+/// A guard to detect whether a future is cancelled.
+pub struct FutureCancelGuard<F: FnMut()> {
+    cancelled: bool,
+    on_cancel: F,
+}
+
+impl<F: FnMut()> FutureCancelGuard<F> {
+    /// Create a guard to assume the future will be cancelled at the following
+    /// await point.
+    ///
+    /// If the future is really cancelled, the provided `on_cancel` callback
+    /// will be executed.
+    pub fn new_cancelled(on_cancel: F) -> Self {
+        Self {
+            cancelled: true,
+            on_cancel,
+        }
+    }
+
+    /// Set the inner state is uncancelled to ensure the `on_cancel` callback
+    /// won't be executed.
+    pub fn uncancel(&mut self) {
+        self.cancelled = false;
+    }
```

Comment:
Personally, I think it feels more appropriate to call it finish, and it looks like the function will only be used when the feature is complete.

Reply:
Sounds reasonable.